From 81e6e9f8ee8accc11f18d1781e6bb962a7a1dcb3 Mon Sep 17 00:00:00 2001
From: Kent Tamura <tkent@chromium.org>
Date: Thu, 06 Jun 2024 02:17:09 +0000
Subject: [PATCH] RubyLB: Fix a crash by an IMG in the alt mode

It contains a floating object.  We should avoid it in rubies.

Bug: 342782844
Change-Id: I625d6a83ed814b3aa1bcba6aeed8fab434a71cb4
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/5601252
Reviewed-by: Koji Ishii <kojii@chromium.org>
Auto-Submit: Kent Tamura <tkent@chromium.org>
Commit-Queue: Koji Ishii <kojii@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1311042}
---

--- a/third_party/blink/renderer/core/html/html_image_fallback_helper.cc
+++ b/third_party/blink/renderer/core/html/html_image_fallback_helper.cc
@@ -189,6 +189,13 @@
 
  private:
   void ShowBrokenImageIcon(ComputedStyleBuilder& builder) {
+    // See AdjustStyleForDisplay() in style_adjuster.cc.
+    if (builder.IsInInlinifyingDisplay()) {
+      builder.SetDisplay(EDisplay::kInline);
+      builder.SetFloating(EFloat::kNone);
+      return;
+    }
+
     // Note that floating elements are blockified by StyleAdjuster.
     builder.SetDisplay(EDisplay::kBlock);
 
--- /dev/null
+++ b/third_party/blink/web_tests/fast/ruby/img-alt-crash.html
@@ -0,0 +1,14 @@
+<!DOCTYPE html>
+<script src="../../resources/testharness.js"></script>
+<script src="../../resources/testharnessreport.js"></script>
+<body>
+<ruby>
+AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA x
+<rt>
+<img alt="AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA AxBxC">
+</ruby>
+<script>
+// crbug.com/342782844
+test(() => {
+}, 'No crash by <img alt> in <ruby>');
+</script>
